### PR TITLE
Add asynchronous communication when creating a track. Closes #45

### LINKED ISSUE #45: Add sending a message to Kafka when creating a new track

### PR DESCRIPTION
**Pull request by DanilByaliy**

### CODE REVIEW

File: services/tracks/src/config/axios.config.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ const axios = require('axios');
 const axiosRetry = require('axios-retry');
 
 const axiosClient = axios.create({
-  baseURL: 'http://host.docker.internal:80/api/v1/',
+  baseURL: 'http://ecstasify-authors-service/api/v1/',
```
Comment (KhrapkoVasyl) [inline services/tracks/src/config/axios.config.js:5]: why not use http://ecstasify/api/v1/authors instead? More general url organization

File: services/tracks/src/services/trackService.js [hunk 2]
```diff
@@ -3,8 +3,26 @@
 const { Tracks } = require('../db/models/Tracks');
 const { v4: uuid } = require('uuid');
 const { axiosClient } = require('../config/axios.config');
+const { Kafka } = require('kafkajs');
+
+const kafka = new Kafka({
```
Comment (KhrapkoVasyl) [inline services/tracks/src/services/trackService.js]: I suggest move the Kafka configuration to a separate file and use the exported configured class instance

File: services/tracks/src/services/trackService.js [hunk 3]
```diff
@@ -43,6 +61,11 @@ class TrackServices {
     const createdTrack = await Tracks.create(trackData);
     createdTrack.author = author;
 
+    await this.notificationProducer.send({
+      topic: 'new-track-created',
```
Comment (KhrapkoVasyl) [inline services/tracks/src/services/trackService.js]: I suggest moving the topic names to a separate file with constants

File: services/tracks/src/services/trackService.js [hunk 4]
```diff
@@ -3,8 +3,26 @@
...
 const { axiosClient } = require('../config/axios.config');
+const { Kafka } = require('kafkajs');
+
+const kafka = new Kafka({
+  clientId: 'my-app',
+  brokers: ['kafka:9092'],
+})
 
 class TrackServices {
+  notificationProducer = kafka.producer();
```
Comment (KhrapkoVasyl) [inline services/tracks/src/services/trackService.js]: Why not make the notificationProducer as a constructor parameter?

File: services/tracks/src/services/trackService.js [hunk 5]
```diff
@@ -3,8 +3,26 @@
...
+
+  constructor() {
+    setTimeout(async () => {
+      try {
+        await this.notificationProducer.connect();
+      } catch (e) {
+        console.error("Error while connecting to kafka: " + e);
+      }
+    }, 60000)
```
Comment (KhrapkoVasyl) [inline services/tracks/src/services/trackService.js]: Why should we use the timeout set here and why is the value 60000?

File: services/tracks/src/services/trackService.js (hunk 2 above)
Comment (KhrapkoVasyl) [reply services/tracks/src/services/trackService.js]: Also I suggest to create a KafkaNotificationProducer class where to implement the required interface

File: services/tracks/src/services/trackService.js [hunk 6]
```diff
@@ -3,8 +3,26 @@
...
 const { v4: uuid } = require('uuid');
 const { axiosClient } = require('../config/axios.config');
+const { Kafka } = require('kafkajs');
+
+const kafka = new Kafka({
+  clientId: 'my-app',
+  brokers: ['kafka:9092'],
+})
 
 class TrackServices {
```
Comment (KhrapkoVasyl) [inline services/tracks/src/services/trackService.js]: If it is possible to rename it in singular? I suggest TrackService would look more logical

**Review by KhrapkoVasyl — CHANGES_REQUESTED** [review]

File: services/tracks/src/config/axios.config.js (hunk 1 above)
Comment (DanilByaliy) [reply services/tracks/src/config/axios.config.js:5]: ecstasify-authors-service is the name of the service we are contacting. We cannot access it through another link.

**Review by KhrapkoVasyl — APPROVED** [review]
Good job!

File: services/tracks/src/controllers/trackController.js [hunk 7]
```diff
@@ -56,7 +56,7 @@ class TrackController {
 
   async postBrockenRequest(req, res) {
     try {
-      const data = await trackServices.postBrockenRequest();
+      const data = await trackService.postBrockenRequest();
```
Comment (artemmatiushenko1) [inline services/tracks/src/controllers/trackController.js]: I guess `postBrockenRequest` method should be called `postBrokenRequest` (typo)

File: services/tracks/src/services/trackService.js [hunk 8]
```diff
@@ -3,8 +3,22 @@
...
 const { v4: uuid } = require('uuid');
 const { axiosClient } = require('../config/axios.config');
+const { kafkaTopics } = require('../constants');
+
+class TrackService {
+  notificationProducer;
+
+  constructor(notificationProducer) {
+    this.notificationProducer = notificationProducer;
+    (async () => {
```
Comment (artemmatiushenko1) [inline services/tracks/src/services/trackService.js]: I would rather make this IIFE as a class method and call it `setup` or smth, looks strange for me

File: services/tracks/src/services/trackService.js [hunk 9]
```diff
@@ -3,8 +3,22 @@
...
+      try {
+        await this.notificationProducer.connect();
+      } catch (e) {
```
Comment (artemmatiushenko1) [inline services/tracks/src/services/trackService.js]: As I saw in your code, you mostly use `err` variable name for errors, so I guess you should change it here

**Review by artemmatiushenko1 — CHANGES_REQUESTED** [review]

**Review by artemmatiushenko1 — APPROVED** [review]

**Review by Bogdan-Zinovij — APPROVED** [review]

**Pull request merged**